### PR TITLE
*: use context to manage runner

### LINKED ISSUE #8386: data race at ConcurrentRunner
### What did you do?
Run PD in a large cluster(by simulator)

### What did you expect to see?
No data race

### What did you see instead?
```bash
==================
WARNING: DATA RACE
Write at 0x00c002e81a48 by goroutine 639:
  github.com/tikv/pd/server/cluster.(*RaftCluster).InitCluster()
      /data/nvme0n1/husharp/proj/pd/server/cluster/cluster.go:299 +0x1184
  github.com/tikv/pd/server/cluster.(*RaftCluster).Start()
      /data/nvme0n1/husharp/proj/pd/server/cluster/cluster.go:313 +0x1dd
  github.com/tikv/pd/server.(*Server).createRaftCluster()
      /data/nvme0n1/husharp/proj/pd/server/server.go:807 +0x8a
  github.com/tikv/pd/server.(*Server).campaignLeader()
      /data/nvme0n1/husharp/proj/pd/server/server.go:1787 +0x187d
  github.com/tikv/pd/server.(*Server).leaderLoop()
      /data/nvme0n1/husharp/proj/pd/server/server.go:1707 +0x1564
  github.com/tikv/pd/server.(*Server).startServerLoop.func1()
      /data/nvme0n1/husharp/proj/pd/server/server.go:664 +0x33

Previous read at 0x00c002e81a48 by goroutine 41370597:
  github.com/tikv/pd/server/cluster.(*RaftCluster).GetRegionStats()
      <autogenerated>:1 +0x3b
  github.com/tikv/pd/pkg/cluster.Collect()
      /data/nvme0n1/husharp/proj/pd/pkg/cluster/cluster.go:81 +0xe1
  github.com/tikv/pd/server/cluster.(*RaftCluster).processRegionHeartbeat.func1()
      /data/nvme0n1/husharp/proj/pd/server/cluster/cluster.go:1066 +0xcb
  github.com/tikv/pd/pkg/ratelimit.(*ConcurrentRunner).run()
      /data/nvme0n1/husharp/proj/pd/pkg/ratelimit/runner.go:149 +0x7a
  github.com/tikv/pd/pkg/ratelimit.(*ConcurrentRunner).Start.func1.2()
      /data/nvme0n1/husharp/proj/pd/pkg/ratelimit/runner.go:121 +0x4f

Goroutine 639 (running) created at:
  github.com/tikv/pd/server.(*Server).startServerLoop()
      /data/nvme0n1/husharp/proj/pd/server/server.go:664 +0x15c
  github.com/tikv/pd/server.(*Server).Run()
      /data/nvme0n1/husharp/proj/pd/server/server.go:636 +0x1ad
  fmt.(*ss).doScanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:1230 +0x3ee
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  github.com/syndtr/goleveldb/leveldb.(*DB).checkAndCleanFiles()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db_util.go:52 +0x23e
  github.com/syndtr/goleveldb/leveldb.openDB()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:130 +0x884
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x224
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:657 +0x16e
  fmt.(*ss).doScanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:1230 +0x3ee
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:476 +0xa6
  github.com/syndtr/goleveldb/leveldb.openDB()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:125 +0x86d
  github.com/syndtr/goleveldb/leveldb.Open()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:197 +0x330
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/session.go:134 +0xbd
  runtime.deferreturn()
      /data/nvme0n1/husharp/go/go1.21.0/src/runtime/panic.go:477 +0x30
  github.com/syndtr/goleveldb/leveldb.Open()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:183 +0xf6
  github.com/syndtr/goleveldb/leveldb.OpenFile()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:219 +0x84
  github.com/tikv/pd/pkg/storage/kv.NewLevelDBKV()
      /data/nvme0n1/husharp/proj/pd/pkg/storage/kv/levedb_kv.go:34 +0x4d
  github.com/tikv/pd/pkg/storage.NewHotRegionsStorage()
      /data/nvme0n1/husharp/proj/pd/pkg/storage/hot_region_storage.go:123 +0x7d
  github.com/tikv/pd/server.(*Server).startServer()
      /data/nvme0n1/husharp/proj/pd/server/server.go:514 +0x2a65
  fmt.(*ss).doScanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:1230 +0x3ee
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  github.com/syndtr/goleveldb/leveldb.(*DB).checkAndCleanFiles()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db_util.go:52 +0x23e
  github.com/syndtr/goleveldb/leveldb.openDB()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:130 +0x884
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x224
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:657 +0x16e
  fmt.(*ss).doScanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:1230 +0x3ee
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:476 +0xa6
  github.com/syndtr/goleveldb/leveldb.openDB()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:125 +0x86d
  github.com/syndtr/goleveldb/leveldb.Open()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:197 +0x330
  fmt.Fscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:143 +0xdc
  fmt.Sscanf()
      /data/nvme0n1/husharp/go/go1.21.0/src/fmt/scan.go:114 +0x164
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:643 +0x1a
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/storage/file_storage.go:458 +0x3f8
  github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/session.go:134 +0xbd
  runtime.deferreturn()
      /data/nvme0n1/husharp/go/go1.21.0/src/runtime/panic.go:477 +0x30
  github.com/syndtr/goleveldb/leveldb.Open()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:183 +0xf6
  github.com/syndtr/goleveldb/leveldb.OpenFile()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20190318030020-c3a204f8e965/leveldb/db.go:219 +0x84
  github.com/tikv/pd/pkg/storage/kv.NewLevelDBKV()
      /data/nvme0n1/husharp/proj/pd/pkg/storage/kv/levedb_kv.go:34 +0x4d
  github.com/tikv/pd/pkg/storage.newLevelDBBackend()
      /data/nvme0n1/husharp/proj/pd/pkg/storage/leveldb_backend.go:62 +0x64
  github.com/tikv/pd/pkg/storage.NewRegionStorageWithLevelDBBackend()
      /data/nvme0n1/husharp/proj/pd/pkg/storage/storage.go:67 +0x64
  github.com/tikv/pd/server.(*Server).startServer()
      /data/nvme0n1/husharp/proj/pd/server/server.go:464 +0xd87
  github.com/tikv/pd/server.(*Server).Run()
      /data/nvme0n1/husharp/proj/pd/server/server.go:627 +0x144
  main.start()
      /data/nvme0n1/husharp/proj/pd/cmd/pd-server/main.go:285 +0x13b1
  main.createServerWrapper()
      /data/nvme0n1/husharp/proj/pd/cmd/pd-server/main.go:191 +0xd5
  github.com/spf13/cobra.(*Command).execute()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0xf9d
  github.com/spf13/cobra.(*Command).ExecuteC()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x604
  github.com/spf13/cobra.(*Command).Execute()
      /data/nvme0n1/husharp/go/go1.21.0/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039 +0x224
  main.main()
      /data/nvme0n1/husharp/proj/pd/cmd/pd-server/main.go:71 +0x225

Goroutine 41370597 (finished) created at:
  github.com/tikv/pd/pkg/ratelimit.(*ConcurrentRunner).Start.func1()
      /data/nvme0n1/husharp/proj/pd/pkg/ratelimit/runner.go:121 +0x36b
==================
```
### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: ref #8386

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by HuSharp — APPROVED** [review]

**Comment by okJiang** [comment]
I found lots of repeated operator logs in the test. Is this because of the content of this pr? 🤔
https://do.pingcap.net/jenkins/blue/rest/organizations/jenkins/pipelines/tikv/pipelines/pd/pipelines/pull_integration_realcluster_test/runs/312/nodes/72/steps/77/log/?start=0
```
[2024/07/15 17:05:31.010 +08:00] [INFO] [operator_controller.go:510] ["add operator"] [region-id=52] [operator="\"grant-leader {transfer leader: store 2 to 3} (kind:leader, region:52(23, 5), createAt:2024-07-15 17:05:31.01069943 +0800 CST m=+78.032028045, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, size:0, steps:[0:{transfer leader from store 2 to store 3}], timeout:[1m0s])\""] [additional-info=]
...
[2024/07/15 17:05:34.512 +08:00] [INFO] [operator_controller.go:510] ["add operator"] [region-id=52] [operator="\"grant-leader {transfer leader: store 2 to 3} (kind:leader, region:52(23, 5), createAt:2024-07-15 17:05:34.512594216 +0800 CST m=+81.533922837, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, size:0, steps:[0:{transfer leader from store 2 to store 3}], timeout:[1m0s])\""] [additional-info=]
...
[2024/07/15 17:05:44.517 +08:00] [INFO] [operator_controller.go:510] ["add operator"] [region-id=52] [operator="\"grant-leader {transfer leader: store 2 to 3} (kind:leader, region:52(23, 5), createAt:2024-07-15 17:05:44.517445078 +0800 CST m=+91.538773699, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, size:0, steps:[0:{transfer leader from store 2 to store 3}], timeout:[1m0s])\""] [additional-info=]
```

**Comment by rleungx** [comment]
> I found lots of repeated operator logs in the test. Is this because of the content of this pr? 🤔 https://do.pingcap.net/jenkins/blue/rest/organizations/jenkins/pipelines/tikv/pipelines/pd/pipelines/pull_integration_realcluster_test/runs/312/nodes/72/steps/77/log/?start=0
> 
> ```
> [2024/07/15 17:05:31.010 +08:00] [INFO] [operator_controller.go:510] ["add operator"] [region-id=52] [operator="\"grant-leader {transfer leader: store 2 to 3} (kind:leader, region:52(23, 5), createAt:2024-07-15 17:05:31.01069943 +0800 CST m=+78.032028045, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, size:0, steps:[0:{transfer leader from store 2 to store 3}], timeout:[1m0s])\""] [additional-info=]
> ...
> [2024/07/15 17:05:34.512 +08:00] [INFO] [operator_controller.go:510] ["add operator"] [region-id=52] [operator="\"grant-leader {transfer leader: store 2 to 3} (kind:leader, region:52(23, 5), createAt:2024-07-15 17:05:34.512594216 +0800 CST m=+81.533922837, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, size:0, steps:[0:{transfer leader from store 2 to store 3}], timeout:[1m0s])\""] [additional-info=]
> ...
> [2024/07/15 17:05:44.517 +08:00] [INFO] [operator_controller.go:510] ["add operator"] [region-id=52] [operator="\"grant-leader {transfer leader: store 2 to 3} (kind:leader, region:52(23, 5), createAt:2024-07-15 17:05:44.517445078 +0800 CST m=+91.538773699, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, size:0, steps:[0:{transfer leader from store 2 to store 3}], timeout:[1m0s])\""] [additional-info=]
> ```

I don't know the relationship between the failure test and this PR. It looks Irrelevant.

File: pkg/ratelimit/runner.go [hunk 1]
```diff
@@ -144,8 +145,13 @@ func (cr *ConcurrentRunner) Start() {
 	}()
 }
 
-func (cr *ConcurrentRunner) run(task *Task, token *TaskToken) {
+func (cr *ConcurrentRunner) run(ctx context.Context, task *Task, token *TaskToken) {
 	start := time.Now()
+	select {
+	case <-ctx.Done():
+		return
```
Comment (nolouch) [inline pkg/ratelimit/runner.go:152]: I think the race problem's main reason is the task of accessing the object without protection. here still cannot avoid that, we should let the task be safely run concurrency in task logic.

Comment (rleungx) [reply pkg/ratelimit/runner.go:152]: Yes, indeed, but with the context, we can exit the goroutine ASAP.

Comment (nolouch) [reply pkg/ratelimit/runner.go:152]: actually, I think it should pass context to the task, and then the task should consider exit ASAP by context cancel. here only prevent starting the new task.  the initial version has this design :)

**Review by nolouch — APPROVED** [review]

**Pull request merged**